### PR TITLE
ONNX export: Square and ReduceSum operators

### PR DESCRIPTION
**Pull request by vandanavk**
## Description ##
Enabling export of Square operator and Sum operator

v2: Added reduce_sum tests based on @anirudhacharya's review
v3: Addressed @Roshrini's and @zhreshold's review comments

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- Export mx.sym.Square using Power operator in ONNX
- Export mx.sym.Sum using ReduceSum operator in ONNX
- Added tests for ReduceSum

## Comments ##
- Tested onnx_backend_test.py

### CODE REVIEW

**Comment by vandanavk** [comment]
@mxnet-label-bot[pr-awaiting-review]

**Comment by vandanavk** [comment]
Rebased on top of #12633.
Added tests for ReduceSum.
Test for square is taken care by the existing tests for Pow operator.

@anirudhacharya @Roshrini @zhreshold This PR is ready for review again.

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 1]
```diff
@@ -2077,3 +2077,85 @@ def convert_sqrt(node, **kwargs):
...
+        return [node]
+    else:
+        node = onnx.helper.make_node(
+            'ReduceSum',
+            inputs=[input_node],
+            outputs=[name],
+            keepdims=keepdims,
+            name=name
+        )
+
```
Comment (anirudhacharya) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: nit: could you remove new line here and in line 2150. Or maybe you could have a single `return` statement after the if else block.

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 2]
```diff
@@ -2077,3 +2077,85 @@ def convert_sqrt(node, **kwargs):
...
+            vals=power2,
+            raw=False,
+        )
+    )
+
+    node = onnx.helper.make_node(
+        "Pow",
+        [input_node_a, power2_name],
+        [name],
+        name=None
```
Comment (anirudhacharya) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: why name=None here?

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 3]
```diff
@@ -2077,3 +2077,85 @@ def convert_sqrt(node, **kwargs):
...
+    onnx = import_onnx_modules()
+    name = node["name"]
+    proc_nodes = kwargs["proc_nodes"]
+    inputs = node["inputs"]
+    attrs = node["attrs"]
+
+    mx_axis = attrs.get("axis", None)
+    axes = convert_string_to_list(str(mx_axis)) if mx_axis is not None else None
+
+    keepdims = 1 if ("keepdims" in attrs) and \
```
[the remark this answers is not in the document]
Comment (vandanavk) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: it's an optional parameter. so added this check just in case.

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 4]
```diff
@@ -2077,3 +2077,85 @@ def convert_sqrt(node, **kwargs):
...
+                   attrs.get("keepdims") in ["True", "1"] else 0
```
[the remark this answers is not in the document]
Comment (vandanavk) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: the value has always been "True" or "1" or "0" when it reaches this function. same is the case with similar functions such as Reduce*, Arg*. I will figure out where this change in type occurs and get back to you.

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py (hunk 1 above)
Comment (vandanavk) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: will change it to a common return statement

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py (hunk 2 above)
Comment (vandanavk) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: was a copy from Pow operator. No reason to have it as None. Will add a name. and change it in Pow as well

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (anirudhacharya) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: no, my point was since you are doing `attrs.get("keepdims")` the first condition is a bit redundant.

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py (hunk 2 above)
Comment (vandanavk) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: Looks like name wasn't added to Pow because of the error "pow() got an unexpected keyword argument 'name'". Submitting the fix for this in this PR itself (separate commit).

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (anirudhacharya) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: I was thinking of this - 
```
keepdims = 1 if attrs.get("keepdims", 0) in ["True", "1"] else 0
```
but this should be ok i guess.

**Comment by Roshrini** [comment]
LGTM

**Comment by vandanavk** [comment]
@anirudhacharya @zhreshold ping for review

File: python/mxnet/contrib/onnx/onnx2mx/_op_translations.py [hunk 5]
```diff
@@ -537,7 +537,8 @@ def power(attrs, inputs, proto_obj):
     if 'broadcast' in attrs and attrs['broadcast'] == 1:
         new_attrs = translation_utils._remove_attributes(new_attrs, ['broadcast'])
         return 'broadcast_power', new_attrs, inputs
-    return 'pow', new_attrs, inputs
+    mxnet_op = symbol.pow(inputs[0], inputs[1])
```
Comment (anirudhacharya) [inline python/mxnet/contrib/onnx/onnx2mx/_op_translations.py]: why this change? you probably made this change to accommodate op_set version 7 of ONNX. will we support op_set versions older than 7?

Comment (vandanavk) [reply python/mxnet/contrib/onnx/onnx2mx/_op_translations.py]: Reverting back to the older commit. Will figure out supporting different op_set versions and work on the Power operator separately.

**Comment by vandanavk** [comment]
@zhreshold @Roshrini @anirudhacharya review comments addressed. is this PR good to go?

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 6]
```diff
@@ -2077,3 +2077,82 @@ def convert_sqrt(node, **kwargs):
...
+    """Map MXNet's square operator attributes to onnx's Pow operator
+    and return the created node.
+    """
+    onnx = import_onnx_modules()
+    name = node["name"]
+    proc_nodes = kwargs["proc_nodes"]
+    inputs = node["inputs"]
+
+    input_node_a_id = kwargs["index_lookup"][inputs[0][0]]
+
```
Comment (anirudhacharya) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: nit: in general feel that there are a lot of unnecessary blank lines in this method. also you can remove line 2096 have `np.array([2])` directly. but this change is not a blocker

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 7]
```diff
@@ -2077,3 +2077,82 @@ def convert_sqrt(node, **kwargs):
...
+    input_node_a = proc_nodes[input_node_a_id].name
+
+    initializer = kwargs["initializer"]
+    power2 = [2]
+    np_arr = np.array(power2)
```
Comment (zhreshold) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: what is the point of setting power2 as a temporary variable

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 8]
```diff
@@ -2077,3 +2077,82 @@ def convert_sqrt(node, **kwargs):
...
+    name = node["name"]
+    proc_nodes = kwargs["proc_nodes"]
+    inputs = node["inputs"]
+    attrs = node["attrs"]
+
+    mx_axis = attrs.get("axis", None)
+    axes = convert_string_to_list(str(mx_axis)) if mx_axis is not None else None
+
+    keepdims = attrs.get("keepdims", 0)
+    keepdims = 1 if keepdims in ["True", "1"] else 0
```
Comment (zhreshold) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: should reuse the function in #12646

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py (hunk 7 above)
Comment (vandanavk) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: Agree. Re-submitting after editing this.

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py (hunk 8 above)
Comment (vandanavk) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: Agree. Re-submitting with this change.

**Comment by vandanavk** [comment]
Since this PR uses changes in https://github.com/apache/incubator-mxnet/pull/12646, combining the 2 PRs. This will be closed in favor of https://github.com/apache/incubator-mxnet/pull/12646 once the changes are pushed.
